src/socket/tcp.rs

```diff
@@ -508,6 +508,26 @@ impl<'a> Socket<'a> {
             panic!("receiving buffer too large, cannot exceed 1 GiB")
         }
         let rx_cap_log2 = mem::size_of::<usize>() * 8 - rx_capacity.leading_zeros() as usize;
+        let remote_win_shift = rx_cap_log2.saturating_sub(16) as u8;
+        Self::new_with_window_scaling(rx_buffer, tx_buffer, remote_win_shift)
+    }
+
+    /// Create a socket using the given buffers and window scaling factor defined in [RFC 1323].
+    ///
+    /// # Panics
+    ///
+    /// Panics if the window scaling factor is greater than 14.
+    ///
+    /// See also the [local_recv_win_scale](#method.local_recv_win_scale) method.
+    pub fn new_with_window_scaling<T>(rx_buffer: T, tx_buffer: T, recv_win_scale: u8) -> Socket<'a>
```

Comment:
By RFC 1323, window scaling can only be set in a SYN segment. Therefore we should force user to set this field in advance rather than give them illusion that it can be changed anytime.

Reply:
That makes sense.
I just think about the future situation where you have another feature that is from the start only, and you want a socket with both features.
A setter which checks for the TCP state (CLOSED/LISTEN) will solve this

Reply:
That makes sense to me. Will update it later.